backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/CachedLibraries.kt

```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright 2010-2019 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license
+ * that can be found in the LICENSE file.
+ */
+
+package org.jetbrains.kotlin.backend.konan
+
+import org.jetbrains.kotlin.konan.file.File
+import org.jetbrains.kotlin.konan.library.KonanLibrary
+import org.jetbrains.kotlin.konan.target.CompilerOutputKind
+import org.jetbrains.kotlin.konan.target.KonanTarget
+import org.jetbrains.kotlin.library.uniqueName
+import org.jetbrains.kotlin.utils.addToStdlib.firstNotNullResult
+
+internal class CachedLibraries(
+        private val target: KonanTarget,
+        allLibraries: List<KonanLibrary>,
+        explicitCaches: Map<KonanLibrary, String>,
+        implicitCacheDirectories: List<String>
+) {
+
+    class Cache(val kind: Kind, val path: String) {
+        enum class Kind { DYNAMIC, STATIC }
+    }
+
+    private val allCaches: Map<KonanLibrary, Cache> = allLibraries.mapNotNull { library ->
+        val explicitPath = explicitCaches[library]
+
+        val cache = if (explicitPath != null) {
+            val kind = when {
+                explicitPath.endsWith(target.family.dynamicSuffix) -> Cache.Kind.DYNAMIC
+                explicitPath.endsWith(target.family.staticSuffix) -> Cache.Kind.STATIC
+                else -> error("unexpected cache: $explicitPath")
```

Comment:
Message can be more helpful. 
Something like "Expected cache file as static ($target.family.staticSuffix) or dynamic (target.family.dynamicSuffix) library. Got: $explicitPath".

Reply:
Also you don't check for file existence in case of explicit path.

Reply:
Note that
1) this is _initial_ support
2) this feature is not entirely visible and recommended for end users directly.
I believe this kinda makes using current message ok.

> Also you don't check for file existence in case of explicit path.

See above. Also linker will check this anyway.